### ISSUE TITLE
Big queue OOM issue

### ISSUE THREAD

**Issue by eawer** [SGrondin/bottleneck#95]
Hi! I'm facing OOM issues while trying to add lot's of jobs to the queue.
Let's say I have a list of 10M links. I need to visit each at allowed 50 requests per sec rate limit.

Here is what I'm doing:

```js
const getData = (url) => {
  const res = await fetch(url)
  ...
}

//I'm not utilizing 50 requests per sec here, but that's not the case
const limiter = new Bottleneck({
  maxConcurrent: 10,
  minTime: 100,
});

for (link of links) {
  limiter.schedule(() => getData(link));
}
```

If `links` contains a considerably small amount of urls, everything works like a charm. But as we remember, `links` is supposed to contain 10M urls. As you may have guessed, I receive `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory` shortly after start (because I'm trying to add 10M of functions (but not the links) to the queue).

Another choice is to do like this:
```
for (link of links) {
  await limiter.schedule(() => getData(link));
}
```
but it kills the whole idea of 50 requests per sec, as it adds the next url only after the previous one was fetched.

Am I doing it wrong? Is there any way to accomplish what I need? Can I just provide a worker function and a list of inputs, so Bottleneck won't be trying to create a huge number of jobs at the same time, but will create a job for each input item only when it's time comes (like here https://caolan.github.io/async/docs.html#queue for example)?

Thanks in advance

**Comment by SGrondin** [comment]
Hi,

Your code will use a lot less memory if you do this instead:
```js
for (link of links) {
  limiter.schedule(getData, link);
}
```
That way JS doesn't need to create a closure and keep 50M copies of your current state in memory.

Please let me know if that fixes it!

**Comment by tcbyrd** [comment]
@SGrondin This is not directly related to this issue, but we're having a similar memory issues with [Probot](https://github.com/probot/probot/) apps after >2.9.0. I looked at the `limiter.on('debug')` messages and can see events that claim it freed the memory, but when I take a heap snapshot I can see that all the objects are still in memory:

<img src="https://user-images.githubusercontent.com/13207348/48985129-a5303b00-f0d2-11e8-854d-f1b74eef782a.png" width="250px" />

This is how we're implementing the rate limiter in probot: 

https://github.com/probot/probot/blob/58559ff22d4524a32fa099db1483fd2c9ac54ca0/src/github/rate-limiting.ts

```js
import Bottleneck from 'bottleneck'
import { GitHubAPI } from './'

export function addRateLimiting (octokit: GitHubAPI, limiter: Bottleneck) {
  if (!limiter) {
    limiter = new Bottleneck({
      maxConcurrent: 1,
      minTime: 1000
    })
  }

  const noop = () => Promise.resolve()
  octokit.hook.before('request', limiter.schedule.bind(limiter, noop))
}
```

Basically it's supposed to schedule a call to GitHub once per second. The memory grows pretty consistently based on how much activity there is on the instance (it grows a lot slower on holidays and weekends):
![image](https://user-images.githubusercontent.com/13207348/48985249-8e3e1880-f0d3-11e8-9fd2-51a793169b46.png)

Any tips here? I tried looking through release notes for 2.9.0, but nothing looked like it should cause this change in behavior.

**Comment by tcbyrd** [comment]
> Have you positively pinpointed v2.9.0 as the beginning of the issue?

We're not 💯on exactly 2.9.0, but we run the Slack integration on Probot and it's showing [v2.8.0](https://github.com/integrations/slack/blob/master/package-lock.json#L951), and that app looks fine. We are specifying `^v2.8.0` and are definitely seeing this in a number of apps that are using versions released after 2.8.0. I can tell you it's definitely happening in v2.10.0.

**Comment by SGrondin** [comment]
Perfect, thank you. I diffed the different versions around 2.9.0 and the most likely culprit is a function scoping change in 2.10.0.

> Awesome! Let me know if I can help with testing.

I'll let you know as soon as I have a branch with the fix.

**Comment by eawer** [comment]
@SGrondin tried your suggestion and unfortunately, it leads to the same issue - `FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory`

**Comment by SGrondin** [comment]
I'm still working on it, the Node debugger segfaults frequently when I take memory snapshots and running the same test twice sometimes yields vastly different results, so progress is slow.

I need to determine:
1. Is it a leak or does the queue just uses more memory in 2.10 than 2.8
2. Which change(s) caused it
3. Why

After 2 short evening sessions on it, I have maybe 50% of the answer to each of these 3 questions. I've identified 3 tiny scoping changes that may be causing it. These changes were made to simplify the code and didn't cause any tests to fail. Once this is fixed I'll definitely be adding a test that validates memory usage with long queues.

@tcbyrd  I'm thinking about pushing 4 branches to Github, one that reverts each of these 3 changes and one that reverts them all. Would you be able to help test them? If not, do you have a reliable script that reproduces the issue? The difficulty I'm having in producing reliable results is slowing me down.

**Comment by tcbyrd** [comment]
> Would you be able to help test them?

I can definitely help test them. The only reliable way I've been able to reproduce the problem is putting load on the server, but we do have a [chaos-monkey](https://github.com/probot/chaos-monkey) app that can reliably send random events to simulate the kind of activity that would cause the leak. This way we can test it outside of something with customers on it.

**Comment by SGrondin** [comment]
Perfect. I'll have those branches up tonight. I'm hoping I can get the hotfix released first, afterwards I can take whatever time I need to find a reliable way to reproduce it locally. Then I'll turn it into a test using a child process.

**Comment by SGrondin** [comment]
Only 2 branches since I was able to fully rule out one of them.
- [scope-fix-1](https://github.com/SGrondin/bottleneck/tree/scope-fix-1)
- [scope-fix-2](https://github.com/SGrondin/bottleneck/tree/scope-fix-2)

Thank you for testing it on your infrastructure! 🎉

**Comment by tcbyrd** [comment]
@SGrondin Thanks! I'm unleashing the chaos moneky on them now! I'll have some data for your tomorrow.

**Comment by tcbyrd** [comment]
So far it looks like both branches are getting GCd as expected:

**2.13.0**:
![image](https://user-images.githubusercontent.com/13207348/49260443-98318580-f40b-11e8-8fca-8b408c240338.png)

**scope-fix-1**:
![image](https://user-images.githubusercontent.com/13207348/49260465-b4352700-f40b-11e8-94a5-5f416bf1f241.png)

**scope-fix-2**:
![image](https://user-images.githubusercontent.com/13207348/49260495-cadb7e00-f40b-11e8-89a2-2074f9a81012.png)

I'm not reading too much into the different shapes of the branches. That graph has a 10 minute resolution and the script generates random loads at random intervals. But it does indicate the objects are getting garbage collected. You can tell in the 2.13.0 branch it just continually increases no matter how much load is applied.

**Comment by SGrondin** [comment]
I've just released v2.13.1, which includes the change in `scope-fix-2`.

Please try it out and let me know if the problem is still occurring.

**Comment by eawer** [comment]
@SGrondin, unfortunately, it is still there. I've created a repo to reproduce the issue - https://github.com/eawer/bottleneck-oom-test. It fails on my machine - Ubuntu 16.04, i7, 16GB

**Comment by SGrondin** [comment]
@tcbyrd Did v2.13.1 remove the need to [pin against v2.9.0](https://github.com/integrations/jira/pull/149)?

@eawer Thank you for the repo, it allowed me really drill down on the issue. It's a different one from @tcbyrd 's. I believe I have identified what's going on and it's not a memory leak: the code is simply not efficient enough. I've narrowed down the main cause to the extensive use of [`apply()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/apply). In my experiments, every `apply` I replaced reduced the noticeably memory usage, and they're all in the hot path. After replacing about 10 of them, I was able to run your test into the millions of queued items. `apply` short-circuits some of V8's optimizations and each use will create a new closure that captures its scope. That was always the intended behavior, and Bottleneck has been using `apply` to launch jobs since the very first version, but newer versions of JS have changed the balance between code clarity and efficiency, and `apply` is no longer on the winning side for a project like this one. Replacing them all isn't very complicated (and has been on my mind for several months), but it will require significant testing/QA, so it's something I'm not going to rush. I'm aiming for "late December" to release this.

**Comment by SGrondin** [comment]
Good news: I finished the `apply()` rewrite, it had little impact on the code and didn't require the amount of testing I expected.

Bad news: It barely reduced the memory usage. In your test it now hits V8's memory usage at around 630k jobs instead of 610k. I have a few ideas on how to hit the gains I saw in my previous tests.

**Comment by Chethansr** [comment]
Is this still an issue with v.2.18, since the bug is still open?

**Comment by SGrondin** [comment]
@Chethansr There's still a limit at around 630k jobs in the queue at the same time, yes. I'm not aware of any memory leak.

Every other month or so I try to find something that could be optimized. I incrementally remove features, change methods into no-ops, etc, and I benchmark the memory usage at every step, to try to narrow it down to one or two places that use too much memory.

It doesn't seem to be any one place in particular, because every `function()`/`() =>` I remove reduces the memory footprint by 1-2%. Functions in JS keep a reference to everything in the scope they were created in. They're "closures" and it's not possible to create non-closure functions in JS. Bottleneck, as you can imagine, contains many nested functions.

I have plans to drastically reorganize the code into something similar to a finite state machine (FSA). The  current code leverages closures and functional programming features whenever possible to stay terse and declarative. A FSA-style refactor would break up all those closures and items in the queue would be holding fewer object references and that would reduce their memory footprint.

**Comment by SGrondin** [comment]
I've done a first pass at refactoring the engine internals to optimize for memory usage.

These numbers are from a 2015 MacBook.

master:
- OOM at 630,000 in queue
- 3.3s to queue 250,000

refactor (step 1):
- OOM at 810,000 in queue
- 2.7s to queue 250,000

More to come.

**Comment by SGrondin** [comment]
refactor (step 3):
- OOM at 1,060,000 in queue
- 2.9s to queue 250,000

I need to evaluate what the theoretical limit would be before I put more time into optimizing this. I highly doubt it's going to reach 10M because certain things are simply unavoidable, like creating a pending Promise for every item, to return to the caller. I'm running out of ideas.

If anybody here could try the `next` branch and report on the memory usage, that would help a lot! All tests are passing, it should behave exactly like master.

**Comment by SGrondin** [comment]
v2.18.1 has been released with these optimizations. It uses 40% less memory per job in the `QUEUED` state. 🎉 

I'm still refactoring to enable new features such as #90 and that may yield more memory savings.

**Comment by luxmeter** [comment]
Hi @SGrondin ,

thanks for you efforts. I love your library! There is a similar issue that I'm facing like the other guys. Maybe you can help me to pin down the problem :)

Basically, I run through a Mongo cursor and hand over an async update function to bottleneck. After 2,8M updates I run into:
```
[INFO ] 20:57:08 - Processed 2839000 documents.

<--- Last few GCs --->

[42500:0x32bd680] 14345828 ms: Scavenge 834.8 (928.0) -> 823.0 (928.0) MB, 3.4 / 0.0 ms  (average mu = 0.994, current mu = 0.995) allocation failure 
[42500:0x32bd680] 14346797 ms: Scavenge 839.3 (928.0) -> 827.2 (928.0) MB, 3.3 / 0.0 ms  (average mu = 0.994, current mu = 0.995) allocation failure 


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x1b03762]
Security context: 0x15f03109a2f1 <JSObject>
    1: _next(aka _next) [0x21aa3b0ff4e9] [/home/ec2-user/foobar/node_modules/bottleneck/lib/Sync.js:~5] [pc=0xb937ca22e32](this=0x17a97fd004d1 <undefined>,0x17a97fd004d1 <undefined>)
    2: /* anonymous */(aka /* anonymous */) [0x21aa3b0fc4a9] [/home/ec2-user/foobar/node_modules/bottleneck/lib/Bottleneck.js:~364] [pc=0xb937ccd0dc... 
```

I have a limit how much I hand over to bottleneck. There should not be more than 200 update requests in flight (queued+running+executing). So I doubt that there are too many objects flying around but I may be wrong.

The loop looks like this:
``` 
    const start = 1
    for await (const [index, line] of enumerate(asyncGenerator, start)) {
        const job = async () => {
            await asyncUpdateFunction(line)
            log.debug(`task finisehd: ${index}`)
            onTaskCompleted(index)
        }
        limiter.schedule(job)
        // fill limiters queue but set a hard limit
        while (jobs() >= maxConcurrentSessions * 2) {
            log.debug(`Too many jobs running or scheduled (${jobs()})`)
            await util.sleep(1000)
        }
    }
``` 

I rather think that I have an issue in my code but I'm not sure. Maybe you know exactly what to do 🙏
